### ISSUE TITLE
Test whether all agda packages build on CI

### ISSUE THREAD

**Issue by turion** [NixOS/nixpkgs#98209]
In https://github.com/NixOS/nixpkgs/pull/98199 I nearly broke agda packages that depend on the agda standard library. It would be nice if there was some automatised check that builds all agda packages. Currently it's not very many packages, so I don't expect this to be a heavy load addition to CI.

I added such a functionality in this commit: https://github.com/NixOS/nixpkgs/pull/98199/commits/e1e2d431fb9fb2e58f9339200a58d5ba54c99e1e But it was out of scope for that PR, and also not a approach in general because that test doesn't concern the agda package set, but agda itself.

In this issue, we should determine and implement whether and how we test whether all agda packages build.

My proposal is adding a small test that builds all agda packages that are not marked as broken.

@alexarice

**Comment by turion** [comment]
@umazalakain @jwiegley Maybe you're interested.

**Comment by turion** [comment]
See https://github.com/NixOS/nixpkgs/pull/98214. Advice welcome.

**Comment by turion** [comment]
It's possible to do this semi-manually in a nix repl with:

`nix-repl> :b agda.withPackages (lib.attrValues (lib.filterAttrs (name: value: (builtins.isAttrs value && builtins.hasAttr "isAgdaDerivation" value && value.isAgdaDerivation && !value.meta.broken)) agdaPackages))`

**Comment by turion** [comment]
I need to know from a CI-knowledgeable person how I would actually trigger a build.

**Comment by turion** [comment]
@alexarice @neosimsim Do you think it makes sense to simply define a derivation `agdaWithAllPackages` and add that as a top-level package? Or as an attribute inside the agda namespace, and then we can build it manually when reviewing?

**Comment by alexarice** [comment]
I think that makes sense, I would suggest adding it in the agda namespace instead of top level, as it's not intended for anyone to use. You probably need to include a removeAttrs to make sure it doesn't try to include itself in it's build.

I believe if this is added to passthru.tests in the agda builder, then it will be automatically built on every agda pr

**Comment by alexarice** [comment]
alternatively this could probably be added as a build input to the nixos test and the nixos test could be called on all agda prs?

**Comment by turion** [comment]
> I believe if this is added to passthru.tests in the agda builder, then it will be automatically built on every agda pr

Where exactly would I add that? In the definition of `withPackages'`? That would be recursive then, though

**Comment by alexarice** [comment]
Think it would go here https://github.com/NixOS/nixpkgs/blob/13a6565c4c5fd34c4dcaf3fc06471a49e100b8ab/pkgs/build-support/agda/default.nix#L75 so that it is put in any package made with `agdaPackages.mkDerivation`

**Comment by turion** [comment]
> alternatively this could probably be added as a build input to the nixos test and the nixos test could be called on all agda prs?

This is what I tried originally in https://github.com/NixOS/nixpkgs/pull/98214. We can still go ahead with that for a while, and when it becomes too much effort to ask ofborg manually all the time, we can look into automatically testing?

**Comment by turion** [comment]
> so that it is put in any package made with `agdaPackages.mkDerivation`

Ah so would you only test in every package individually whether it could be built?

**Comment by alexarice** [comment]
> This is what I tried originally in #98214. We can still go ahead with that for a while, and when it becomes too much effort to ask ofborg manually all the time, we can look into automatically testing?

I just reread that thread and now think it would be a bit weird to put this in the agda infrastructure test

**Comment by alexarice** [comment]
> > so that it is put in any package made with `agdaPackages.mkDerivation`
> 
> Ah so would you only test in every package individually whether it could be built?

The tests for every package would test if every other package compiled.

**Comment by turion** [comment]
> I just reread that thread and now think it would be a bit weird to put this in the agda infrastructure test

I can also pull this part out of the nixos test and _only_ have it in `passthru.tests`. Is that good?

> > > so that it is put in any package made with `agdaPackages.mkDerivation`
> > 
> > 
> > Ah so would you only test in every package individually whether it could be built?
> 
> The tests for every package would test if every other package compiled.

Which is fine because compiled packages are cached, I guess?

**Comment by turion** [comment]
> > The tests for every package would test if every other package compiled.
> 
> Which is fine because compiled packages are cached, I guess?

The tests are not very specific then, though. If one library breaks, then _all_ agda packages' tests fail. It would be cool if the test would be like `agdaWithPackages [ self ]`, but that would need some recursion somewhere, and I'm not sure where.

**Comment by alexarice** [comment]
> > > The tests for every package would test if every other package compiled.
> > 
> > 
> > Which is fine because compiled packages are cached, I guess?
> 
> The tests are not very specific then, though. If one library breaks, then _all_ agda packages' tests fail. It would be cool if the test would be like `agdaWithPackages [ self ]`, but that would need some recursion somewhere, and I'm not sure where.

You could change the filter to only include packages which have the package itself as a build input

**Comment by neosimsim** [comment]
> @alexarice @neosimsim Do you think it makes sense to simply define a derivation `agdaWithAllPackages` and add that as a top-level package? Or as an attribute inside the agda namespace, and then we can build it manually when reviewing?

Would that even be possible if packages have colliding modules? Should be right, since every package is built with its own `agda.withPackages`?

**Comment by alexarice** [comment]
> Would that even be possible if packages have colliding modules? Should be right, since every package is built with its own `agda.withPackages`?
Agda only complains when you actually try to use one of the colliding modules as far as I know.

**Comment by neosimsim** [comment]
> > Would that even be possible if packages have colliding modules? Should be right, since every package is built with its own `agda.withPackages`?
> 
> Agda only complains when you actually try to use one of the colliding modules as far as I know.

That's how I understand it, too. But what I was trying to say is, that `agda.withAllPackages` wouldn't even fail because of ambiguous modules, unless a package, e.g. `cubical`, would already fail.

**Comment by alexarice** [comment]
Yes, `agda.withAllPackages` should fail if and only if one of the input packages fails

**Comment by neosimsim** [comment]
I just wonder if there would be a reason to ever not use `agda.withAllPackages`?

**Comment by turion** [comment]
> I just wonder if there would be a reason to ever not use `agda.withAllPackages`?

> The tests are not very specific then, though. If one library breaks, then all agda packages' tests fail.

**Comment by turion** [comment]
This is actively being worked on in https://github.com/NixOS/nixpkgs/pull/98214
